src/automata.jl

```diff
@@ -18,38 +18,54 @@ constitute `AbstractState` interface:
  * `outedges`: gives the vector (indexed by position of the letter in the alphabet)
     of states to which there is an edge starting at given state and labelled by
     the given letter
+ * `isnoedge`: returns whether the state represents lack of edge or not
  * `Base.length`: gives the length of the signature of the shortest path to the
     state, which starts at initial state (length of "simple path")
 """
-abstract type AbstractState{T} end
+abstract type AbstractState{T, N} end
 
 """
-    State{T} <: AbstractState{T}
+    State{T, N} <: AbstractState{T, N}
 State as a name (corresponding to the signature of the simple path ending at this
 state), indication whether it is terminal (in which case there is a right-hand part
 of the rewriting rule, vector of states (indexed by the position of the letter in
 the alphabet) from which there is an edge (labelled by letter indexed) to this state
 and vector of states (indexed by position of the letter in the alphabet) to which
 there is an edge (labelled by letter indexed) from this state.
 """
-mutable struct State{T} <: AbstractState{T}
-    name::AbstractWord{T}
+mutable struct State{T, N} <: AbstractState{T, N}
+    name::Word{T}
     terminal::Bool
-    rrule::AbstractWord{T}
-    ined::Vector{Union{State{T}, Nothing}}
-    outed::Vector{Union{State{T}, Nothing}}
-end
-
+    rrule::Word{T}
+    ined::Vector{State{T, N}}
+    outed::NTuple{N, State{T, N}}
+    representsnoedge::Bool
+
+    function State{T, N}() where {T, N}
+        x = new()
+        x.representsnoedge = true
+        return x
+    end
```

Comment:
please parametrize `State{T,N,W<:AbstractWord{T}} <: AbstractState{T,N}` by the type of the word.
The parameter needs to be added to the inner constructors then as well (to avoid abstract types).

Reply:
So I `btime`'ed it and what you propose is slower:

```julia
julia> let R = RWS_Example_237_abaB(4), seed=1
                  rws = KnuthBendix.knuthbendix2(R)

                  Random.seed!(seed)
                  n = length(rws.order.A.alphabet)
                  w = Word(rand(1:n, 2000))

                  @btime KnuthBendix.rewrite_from_left($w, $rws)
                  ia = KnuthBendix.makeindexautomaton(rws, KnuthBendix.ordering(rws).A)
                  @btime KnuthBendix.index_rewrite($w, $ia)
               end;
  6.926 ms (543115 allocations: 17.33 MiB)
  1.212 ms (6050 allocations: 280.06 KiB)
```

vs the one I did:

```julia
julia> let R = RWS_Example_237_abaB(4), seed=1
           rws = KnuthBendix.knuthbendix2(R)

           Random.seed!(seed)
           n = length(rws.order.A.alphabet)
           w = Word(rand(1:n, 2000))

           @btime KnuthBendix.rewrite_from_left($w, $rws)
           ia = KnuthBendix.makeindexautomaton(rws, KnuthBendix.ordering(rws).A)
           @btime KnuthBendix.index_rewrite($w, $ia)
       end;
  6.933 ms (543115 allocations: 17.33 MiB)
  501.837 μs (4594 allocations: 234.56 KiB)
```

Although I am not sure if I implemented your suggestion correctly:
```julia
mutable struct State{T, N, W<:AbstractWord{T}} <: AbstractState{T, N, W}
    name::W
    terminal::Bool
    rrule::W
    ined::Vector{State{T, N, W}}
    outed::NTuple{N, State{T, N, W}}
    representsnoedge::Bool

    function State{T, N, W}() where {T, N, W}
        x = new()
        x.representsnoedge = true
        return x
    end

    function State(name::W, absize::Integer, noedge::State{T, N, W}) where {T, N, W}
        @assert N == absize "Length of the alphabet does not match!"
        s = State{T, N, W}()
        s.name = name
        s.terminal = false
        s.rrule = W()
        s.ined = State{T, absize}[]
        s.outed = ntuple(i -> noedge, absize)
        s.representsnoedge = false
        return s
    end
end
```

Reply:
most probably there is something abstractly typed now in the code:
```
1.212 ms (6050 allocations: 280.06 KiB)
```
vs
```
501.837 μs (4594 allocations: 234.56 KiB)
```
adding parameter doesn't produce new allocations

Reply:
```
s.ined = State{T, absize}[]
```
is abstractly typed, you need `State{T, N, W}[]`

Reply:
It required slightly more work but here it is:

```julia
julia> let R = RWS_Example_237_abaB(4), seed=1
           rws = KnuthBendix.knuthbendix2(R)

           Random.seed!(seed)
           n = length(rws.order.A.alphabet)
           w = Word(rand(1:n, 2000))

           @btime KnuthBendix.rewrite_from_left($w, $rws)
           ia = KnuthBendix.makeindexautomaton(rws, KnuthBendix.ordering(rws).A)
           @btime KnuthBendix.index_rewrite($w, $ia)
       end;
  6.877 ms (543115 allocations: 17.33 MiB)
  218.877 μs (4594 allocations: 234.56 KiB)
```